bin/varnishd/cache/cache_director.c

```diff
@@ -103,24 +103,26 @@ VRT_VDI_Resolve(VRT_CTX, VCL_BACKEND d)
 static VCL_BACKEND
 VDI_Resolve(VRT_CTX)
 {
-	const struct director *d;
+	VCL_BACKEND d;
 	struct busyobj *bo;
 
+	CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);
 	bo = ctx->bo;
 	CHECK_OBJ_NOTNULL(bo, BUSYOBJ_MAGIC);
-	CHECK_OBJ_ORNULL(bo->director_req, DIRECTOR_MAGIC);
-	d = VRT_VDI_Resolve(ctx, bo->director_req);
-	CHECK_OBJ_ORNULL(d, DIRECTOR_MAGIC);
-	if (d == NULL) {
-		VSLb(bo->vsl, SLT_FetchError, "No backend");
-	} else {
-		AN(d->vdir);
 
-		if (d->vdir->admin_health->health == 0)
-			d = NULL;
+	if (bo->director_req == NULL) {
+		VSLb(bo->vsl, SLT_FetchError, "No backend");
+		return (NULL);
 	}
 
-	return (d);
+	CHECK_OBJ(bo->director_req, DIRECTOR_MAGIC);
+	d = VRT_VDI_Resolve(ctx, bo->director_req);
```

Comment:
`VRT_VDI_Resolve()` is not part of any release yet, should we enforce that callers provide a non-null director?

Reply:
Since we are following #2680 up, I liked `VRT_DirectorResolve()` better than `VRT_VDI_Resolve()`.

Reply:
> `VRT_VDI_Resolve()` is not part of any release yet, should we enforce that callers provide a non-null director?

`NULL` is now a perfectly valid backend (called `none` in VCL) and I do not see why we should add an assertion. Do you have any specific scenario in mind?

Reply:
> `NULL` is now a perfectly valid backend (called `none` in VCL) and I do not see why we should add an assertion.

I forgot.

> Do you have any specific scenario in mind?

No, it was more a case of whether it was considered. I forgot that legit VCL can yield a null backend.